OnPageCode.js

```diff
@@ -1,219 +1,411 @@
-    if (userAgent.match(/opr\//i)) return "opera";
-    if (userAgent.match(/edg/i)) return "edge";
-    return "others";
-  })(),
-
-  // to make sure that we do not read the stats for the same network call again
-  'lastExecutionMaxIndex': 0,
-
-  getTimeValue: function (endTime, startTime) {
-    if (isNaN(endTime) || isNaN(startTime)) return -1;
-    if (endTime === 0 || startTime === 0) return -1;
-    return endTime - startTime;
-  },
-
-  uploadTheNetworkLatencyData: function (jsonData) {
-    fetch("https://pm-network-latency-monitoring.harshad.workers.dev/", {
-      "headers": {
-        "content-type": "application/json"
-      },
-      "method": "POST",
-      "body": JSON.stringify(jsonData)
-    })
-      .then(res => { })
-      .then(response => { })
-      .catch(error => { });
-  },
-
-  getParameterByName: function (name, url) {
-    name = name.replace(/[\[\]]/g, '\\$&');
-    var regex = new RegExp('[?&]' + name + '(=([^&#]*)|&|#|$)'),
-      results = regex.exec(url);
-    if (!results)
-      return null;
-    if (!results[2])
-      return '';
-    return decodeURIComponent(results[2].replace(/\+/g, ' '));
-  },
-
-  prepareNetworkLatencyData: function (perfResource, sspConfig, timelines) {
-    let output = {
-      domain: PM_Network_POC.domain,
-      publisherId: PM_Network_POC.publisherId,
-      browser: PM_Network_POC.browserName,
-      timestamp: Date.now(),
-      bidder: {},
-      nw: {
-        evaluated: {},
+
+    'bidCallTracker': [],
+
+    prepareNetworkLatencyData: function (perfResource, sspConfig, bidderRequest, latency, processedBidsCount) {
+      let currentBidCall = {};
+      let output = {
+        domain: PM_Network_POC.domain,
+        publisherId: PM_Network_POC.publisherId,
+        browser: PM_Network_POC.browserName,
+        platform: PM_Network_POC.platform,
+        adUnitCount: PM_Network_POC.adUnitCount,
+        atLeastOneBidResUsedInAuction: sspConfig.atLeastOneBidResUsedInAuction,
+        bidder: sspConfig.bidderCode,
+        bidCalls: PM_Network_POC.bidCallTracker,
+        serverLatency: latency || {},
+        t: PM_Network_POC.timeoutCorrelators[bidderRequest?.nwMonitor?.correlator] ? 1 : 0,
+        db: perfResource?.name?.length ? 0 : 1
+      };
+
+      if (PM_Network_POC.networkType) output['networkType'] = PM_Network_POC.networkType;
+
+      currentBidCall.order = sspConfig.bids[processedBidsCount - 1].order;
+      currentBidCall.reqMethod = sspConfig.bids[processedBidsCount - 1].reqMethod;
+      currentBidCall.timestamp = Date.now();
+      currentBidCall.nw = {
+        evaluated: {
+          // percieved latency
+          per_lt: Date.now() - PM_Network_POC.cache.auction.timestamp
+        },
         raw: {}
-      },
-      serverTimelines: timelines
-    };
+      };
+      currentBidCall.requestUrlPayloadLength = bidderRequest?.nwMonitor?.requestUrlPayloadLength || sspConfig.bids[processedBidsCount - 1].reqPayloadCharCount;
```

Comment:
Is `bidderRequest?.nwMonitor?.requestUrlPayloadLength` still needed now that we can dynamically access the payload length from pbjs? i left the `reqPayloadCharCount` taken from pbjs as a fallback.

Reply:
Yes, because I don't think so, this payload length from pbjs has the length of the URL included?? whereas `bidderRequest?.nwMonitor?.requestUrlPayloadLength` includes URL length.

Reply:
reqPayloadCharCount just includes the payload we are passing

```
"{"id":"1673035265635","at":1,"cur":["USD"],"imp":[{"id":"2661c0a13025c8","tagid":"pubmatic_test","secure":1,"ext":{"data":{"pbadslot":"div-gpt-ad-1460505748561-0"}},"bidfloorcur":"USD","banner":{"w":300,"h":250,"format":[{"w":300,"h":600}],"pos":0,"topframe":1}}],"site":{"page":"http://localhost:9999/integrationExamples/gpt/hello_world.html","ref":"http://localhost:9999/integrationExamples/gpt/hello_world.html","publisher":{"id":"156276","domain":"localhost:9999"},"domain":"localhost"},"device":{"ua":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36","js":1,"dnt":0,"h":1440,"w":3440,"language":"en","geo":{}},"user":{"geo":{}},"ext":{"wrapper":{"wiid":"de738bf7-6ea2-43b7-a7bd-f53ddb27152b","wv":"prebid_prebid_7.32.0-pre","transactionId":"f1b50d55-90e1-4941-8284-592411098aed","wp":"pbjs"},"epoch":1673035265636},"source":{"tid":"f1b50d55-90e1-4941-8284-592411098aed"},"tmax":1000}"
```

bidderRequest?.nwMonitor?.requestUrlPayloadLength includes the pm url and payload?

Reply:
maybe `bidderRequest?.nwMonitor?.requestUrlPayloadLength` and `reqPayloadCharCount` should be captured and treated as separate fields?

Reply:
Yes, bidderRequest?.nwMonitor?.requestUrlPayloadLength includes the pm URL and payload.